<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Identify Spatial Homogeneous Regions with concordex • concordexR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Identify Spatial Homogeneous Regions with concordex">
<meta name="description" content="Spatial homogeneous regions (SHRs) in tissues are domains that are homogenous with respect to cell type composition. We present a method for identifying SHRs using spatial transcriptomics data, and demonstrate that it is efficient and effective at finding SHRs for a wide variety of tissue types. concordex relies on analysis of k-nearest-neighbor (kNN) graphs. The tool is also useful for analysis of non-spatial transcriptomics data, and can elucidate the extent of concordance between partitions of cells derived from clustering algorithms, and transcriptomic similarity as represented in kNN graphs.">
<meta property="og:description" content="Spatial homogeneous regions (SHRs) in tissues are domains that are homogenous with respect to cell type composition. We present a method for identifying SHRs using spatial transcriptomics data, and demonstrate that it is efficient and effective at finding SHRs for a wide variety of tissue types. concordex relies on analysis of k-nearest-neighbor (kNN) graphs. The tool is also useful for analysis of non-spatial transcriptomics data, and can elucidate the extent of concordance between partitions of cells derived from clustering algorithms, and transcriptomic similarity as represented in kNN graphs.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">concordexR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/concordex-nonspatial.html">Using concordex in to assess cluster boundaries in scRNA-seq</a></li>
    <li><a class="dropdown-item" href="articles/overview.html">Overview of concordexR</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/concordexR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="concordexr">concordexR<a class="anchor" aria-label="anchor" href="#concordexr"></a>
</h1></div>
<!-- badges: start -->

<p>The goal of concordexR is to identify spatial homogeneous regions (SHRs) as defined in the recent manuscript, <a href="https://doi.org/10.1101/2023.06.28.546949" class="external-link">“Identification of spatial homogenous regions in tissues with concordex”</a>. Briefly, SHRs are are domains that are homogeneous with respect to cell type composition. <code>concordex</code> relies on the the k-nearest-neighbor (kNN) graph to representing similarities between cells and uses common clustering algorithms to identify SHRs.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>(Recommended) You can install the development version of <code>concordexR</code> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pachterlab/concordexR"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using GitHub PAT from the git credential store.</span></span>
<span><span class="co">#&gt; Downloading GitHub repo pachterlab/concordexR@HEAD</span></span>
<span><span class="co">#&gt; matrixStats (1.4.1    -&gt; 1.5.0   ) [CRAN]</span></span>
<span><span class="co">#&gt; bit         (4.5.0    -&gt; 4.5.0.1 ) [CRAN]</span></span>
<span><span class="co">#&gt; cpp11       (0.5.0    -&gt; 0.5.1   ) [CRAN]</span></span>
<span><span class="co">#&gt; bit64       (4.5.2    -&gt; 4.6.0-1 ) [CRAN]</span></span>
<span><span class="co">#&gt; pillar      (1.9.0    -&gt; 1.10.1  ) [CRAN]</span></span>
<span><span class="co">#&gt; RSQLite     (2.3.7    -&gt; 2.3.9   ) [CRAN]</span></span>
<span><span class="co">#&gt; Rcpp        (1.0.13-1 -&gt; 1.0.14  ) [CRAN]</span></span>
<span><span class="co">#&gt; BH          (1.84.0-0 -&gt; 1.87.0-1) [CRAN]</span></span>
<span><span class="co">#&gt; igraph      (2.1.1    -&gt; 2.1.3   ) [CRAN]</span></span>
<span><span class="co">#&gt; Skipping 21 packages ahead of CRAN: zlibbioc, XVector, SparseArray, S4Arrays, IRanges, S4Vectors, MatrixGenerics, BiocGenerics, GenomeInfoDbData, GenomeInfoDb, DelayedArray, Biobase, GenomicRanges, SummarizedExperiment, BiocParallel, BiocFileCache, SingleCellExperiment, BiocNeighbors, SpatialExperiment, sparseMatrixStats, bluster</span></span>
<span><span class="co">#&gt; Installing 9 packages: matrixStats, bit, cpp11, bit64, pillar, RSQLite, Rcpp, BH, igraph</span></span>
<span><span class="co">#&gt; Installing packages into '/private/var/folders/xt/1wqk38p91gj1rj16t201p9nw0000gn/T/RtmpaxG0sT/temp_libpath5bb940deed66'</span></span>
<span><span class="co">#&gt; (as 'lib' is unspecified)</span></span>
<span><span class="co">#&gt; Warning: unable to access index for repository https://bioconductor.org/packages/3.18/bioc/bin/macosx/big-sur-x86_64/contrib/4.4:</span></span>
<span><span class="co">#&gt;   cannot open URL 'https://bioconductor.org/packages/3.18/bioc/bin/macosx/big-sur-x86_64/contrib/4.4/PACKAGES'</span></span>
<span><span class="co">#&gt; Warning: unable to access index for repository https://bioconductor.org/packages/3.18/data/annotation/bin/macosx/big-sur-x86_64/contrib/4.4:</span></span>
<span><span class="co">#&gt;   cannot open URL 'https://bioconductor.org/packages/3.18/data/annotation/bin/macosx/big-sur-x86_64/contrib/4.4/PACKAGES'</span></span>
<span><span class="co">#&gt; Warning: unable to access index for repository https://bioconductor.org/packages/3.18/data/experiment/bin/macosx/big-sur-x86_64/contrib/4.4:</span></span>
<span><span class="co">#&gt;   cannot open URL 'https://bioconductor.org/packages/3.18/data/experiment/bin/macosx/big-sur-x86_64/contrib/4.4/PACKAGES'</span></span>
<span><span class="co">#&gt; Warning: unable to access index for repository https://bioconductor.org/packages/3.18/workflows/bin/macosx/big-sur-x86_64/contrib/4.4:</span></span>
<span><span class="co">#&gt;   cannot open URL 'https://bioconductor.org/packages/3.18/workflows/bin/macosx/big-sur-x86_64/contrib/4.4/PACKAGES'</span></span>
<span><span class="co">#&gt; Warning: unable to access index for repository https://bioconductor.org/packages/3.18/books/bin/macosx/big-sur-x86_64/contrib/4.4:</span></span>
<span><span class="co">#&gt;   cannot open URL 'https://bioconductor.org/packages/3.18/books/bin/macosx/big-sur-x86_64/contrib/4.4/PACKAGES'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The downloaded binary packages are in</span></span>
<span><span class="co">#&gt;  /var/folders/xt/1wqk38p91gj1rj16t201p9nw0000gn/T//RtmpzKfXwP/downloaded_packages</span></span>
<span><span class="co">#&gt; ── R CMD build ─────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">#&gt; * checking for file ‘/private/var/folders/xt/1wqk38p91gj1rj16t201p9nw0000gn/T/RtmpzKfXwP/remotes5ce53aa7effb/pachterlab-concordexR-1d49855/DESCRIPTION’ ... OK</span></span>
<span><span class="co">#&gt; * preparing ‘concordexR’:</span></span>
<span><span class="co">#&gt; * checking DESCRIPTION meta-information ... OK</span></span>
<span><span class="co">#&gt; * checking for LF line-endings in source and make files and shell scripts</span></span>
<span><span class="co">#&gt; * checking for empty or unneeded directories</span></span>
<span><span class="co">#&gt; * building ‘concordexR_1.7.1.tar.gz’</span></span>
<span><span class="co">#&gt; Installing package into '/private/var/folders/xt/1wqk38p91gj1rj16t201p9nw0000gn/T/RtmpaxG0sT/temp_libpath5bb940deed66'</span></span>
<span><span class="co">#&gt; (as 'lib' is unspecified)</span></span></code></pre></div>
<p>Versions of the <code>concordexR</code> package that do not enable clustering spatial data into spatial homogeneous regions (SHRs) are available for Bioconductor versions 3.17-19. The most recent version of the package is slated to be released on Bioconductor version 3.20. You can install the Bioconductor development version with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">"3.20"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'getOption("repos")' replaces Bioconductor standard repositories, see</span></span>
<span><span class="co">#&gt; 'help("repositories", package = "BiocManager")' for details.</span></span>
<span><span class="co">#&gt; Replacement repositories:</span></span>
<span><span class="co">#&gt;     CRAN: https://cran.rstudio.com/</span></span>
<span><span class="co">#&gt; Bioconductor version 3.20 (BiocManager 1.30.25), R 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Old packages: 'bit64', 'igraph', 'anytime', 'arrow', 'assorthead', 'BH',</span></span>
<span><span class="co">#&gt;   'BiocCheck', 'BiocNeighbors', 'Biostrings', 'bit', 'bookdown', 'class',</span></span>
<span><span class="co">#&gt;   'classInt', 'clue', 'cluster', 'cpp11', 'data.table', 'DelayedMatrixStats',</span></span>
<span><span class="co">#&gt;   'edgeR', 'evaluate', 'fitdistrplus', 'fontawesome', 'foreign',</span></span>
<span><span class="co">#&gt;   'GenomeInfoDb', 'geometry', 'ggmapinset', 'graph', 'graphlayouts', 'IRanges',</span></span>
<span><span class="co">#&gt;   'KernSmooth', 'knitr', 'later', 'limma', 'lme4', 'lpSolve', 'lubridate',</span></span>
<span><span class="co">#&gt;   'maps', 'MASS', 'MatrixGenerics', 'matrixStats', 'nnet', 'parallelly',</span></span>
<span><span class="co">#&gt;   'pillar', 'pkgbuild', 'processx', 'progressr', 'promises', 'quantreg',</span></span>
<span><span class="co">#&gt;   'Rcpp', 'RcppArmadillo', 'Rdpack', 'reticulate', 'Rfast', 'rhdf5', 'rpart',</span></span>
<span><span class="co">#&gt;   'RSQLite', 'Seurat', 'SFEData', 'shiny', 'SingleCellExperiment', 'spatial',</span></span>
<span><span class="co">#&gt;   'SpatialFeatureExperiment', 'spatialLIBD', 'spatstat.data',</span></span>
<span><span class="co">#&gt;   'spatstat.explore', 'spatstat.geom', 'spatstat.utils', 'spData', 'spdep',</span></span>
<span><span class="co">#&gt;   'stringdist', 'survival', 'systemfonts', 'terra', 'testthat', 'textshaping',</span></span>
<span><span class="co">#&gt;   'tidyterra', 'waldo', 'xfun', 'XML'</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"concordexR"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'getOption("repos")' replaces Bioconductor standard repositories, see</span></span>
<span><span class="co">#&gt; 'help("repositories", package = "BiocManager")' for details.</span></span>
<span><span class="co">#&gt; Replacement repositories:</span></span>
<span><span class="co">#&gt;     CRAN: https://cran.rstudio.com/</span></span>
<span><span class="co">#&gt; Bioconductor version 3.20 (BiocManager 1.30.25), R 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Warning: package(s) not installed when version(s) same as or greater than current; use</span></span>
<span><span class="co">#&gt;   `force = TRUE` to re-install: 'concordexR'</span></span>
<span><span class="co">#&gt; Old packages: 'bit64', 'igraph', 'anytime', 'arrow', 'assorthead', 'BH',</span></span>
<span><span class="co">#&gt;   'BiocCheck', 'BiocNeighbors', 'Biostrings', 'bit', 'bookdown', 'class',</span></span>
<span><span class="co">#&gt;   'classInt', 'clue', 'cluster', 'cpp11', 'data.table', 'DelayedMatrixStats',</span></span>
<span><span class="co">#&gt;   'edgeR', 'evaluate', 'fitdistrplus', 'fontawesome', 'foreign',</span></span>
<span><span class="co">#&gt;   'GenomeInfoDb', 'geometry', 'ggmapinset', 'graph', 'graphlayouts', 'IRanges',</span></span>
<span><span class="co">#&gt;   'KernSmooth', 'knitr', 'later', 'limma', 'lme4', 'lpSolve', 'lubridate',</span></span>
<span><span class="co">#&gt;   'maps', 'MASS', 'MatrixGenerics', 'matrixStats', 'nnet', 'parallelly',</span></span>
<span><span class="co">#&gt;   'pillar', 'pkgbuild', 'processx', 'progressr', 'promises', 'quantreg',</span></span>
<span><span class="co">#&gt;   'Rcpp', 'RcppArmadillo', 'Rdpack', 'reticulate', 'Rfast', 'rhdf5', 'rpart',</span></span>
<span><span class="co">#&gt;   'RSQLite', 'Seurat', 'SFEData', 'shiny', 'SingleCellExperiment', 'spatial',</span></span>
<span><span class="co">#&gt;   'SpatialFeatureExperiment', 'spatialLIBD', 'spatstat.data',</span></span>
<span><span class="co">#&gt;   'spatstat.explore', 'spatstat.geom', 'spatstat.utils', 'spData', 'spdep',</span></span>
<span><span class="co">#&gt;   'stringdist', 'survival', 'systemfonts', 'terra', 'testthat', 'textshaping',</span></span>
<span><span class="co">#&gt;   'tidyterra', 'waldo', 'xfun', 'XML'</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>This is a basic example using concordex:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Voyager</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SpatialFeatureExperiment</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SpatialFeatureExperiment'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     scale</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     saveRDS</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/concordexR" class="external-link">concordexR</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"small"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">quickCluster</span><span class="op">(</span><span class="va">sfe</span>, min.size<span class="op">=</span><span class="fl">2</span>, d<span class="op">=</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calculateConcordex.html">calculateConcordex</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span>,</span>
<span>    <span class="va">clusters</span>,</span>
<span>    n_neighbors<span class="op">=</span><span class="fl">10</span>,</span>
<span>    BLUSPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/KmeansParam-class.html" class="external-link">KmeansParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="st">"shr"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">nbc</span>, <span class="st">"shrs"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotSpatialFeature</span><span class="op">(</span><span class="va">sfe</span>, features<span class="op">=</span><span class="st">"shr"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-plot-shr-1.png" width="70%"></p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you’d like to use the <code>concordexR</code> package in your research, please cite our recent bioRxiv preprint</p>
<blockquote>
<p>Jackson, K.; Booeshaghi, A. S.; Gálvez-Merchán, Á.; Moses, L.; Chari, T.; Pachter, L. Quantitative assessment of single-cell RNA-seq clustering with CONCORDEX. bioRxiv (Cold Spring Harbor Laboratory) 2023. <a href="https://doi.org/10.1101/2023.06.28.546949" class="external-link uri">https://doi.org/10.1101/2023.06.28.546949</a>.</p>
</blockquote>
<p>@article {Jackson2023.06.28.546949, author = {Jackson, Kayla C. and Booeshaghi, A. Sina and G{’a}lvez-Merch{’a}n, {’A}ngel and Moses, Lambda and Chari, Tara and Kim, Alexandra and Pachter, Lior}, title = {Identification of spatial homogeneous regions in tissues with concordex}, year = {2024}, doi = {10.1101/2023.06.28.546949}, publisher = {Cold Spring Harbor Laboratory}, URL = {<a href="https://www.biorxiv.org/content/early/2024/07/18/2023.06.28.546949" class="external-link uri">https://www.biorxiv.org/content/early/2024/07/18/2023.06.28.546949</a>}, journal = {bioRxiv} }</p>
</div>
<div class="section level2">
<h2 id="maintainer">Maintainer<a class="anchor" aria-label="anchor" href="#maintainer"></a>
</h2>
<p><a href="https://github.com/kayla-jackson" class="external-link">Kayla Jackson</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/concordexR" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/pachterlab/concordexR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/pachterlab/concordexR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing concordexR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kayla Jackson <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-6483-0108" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>A. Sina Booeshaghi <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6442-4502" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Angel Galvez-Merchan <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7420-8697" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lambda Moses <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7092-9427" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lior Pachter <br><small class="roles"> Author, reviewer, thesis advisor </small> <a href="https://orcid.org/0000-0002-9164-6231" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://codecov.io/gh/pachterlab/concordexR" class="external-link"><img src="https://codecov.io/gh/pachterlab/concordexR/branch/main/graph/badge.svg?token=FSASJPR4T5" alt="codecov"></a></li>
<li><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/concordexR" class="external-link"><img src="http://www.bioconductor.org/shields/build/devel/bioc/concordexR.svg" alt="Bioc devel status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kayla Jackson, A. Sina Booeshaghi, Angel Galvez-Merchan, Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
