<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using concordex in to assess cluster boundaries in scRNA-seq • concordexR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using concordex in to assess cluster boundaries in scRNA-seq">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">concordexR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/concordex-nonspatial.html">Using concordex in to assess cluster boundaries in scRNA-seq</a></li>
    <li><a class="dropdown-item" href="../articles/overview.html">Overview of concordexR</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/concordexR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Using concordex in to assess cluster boundaries in scRNA-seq</h1>
                        <h4 data-toc-skip class="author">Lambda Moses,
Kayla Jackson</h4>
            
            <h4 data-toc-skip class="date">Jul 22, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/concordexR/blob/main/vignettes/concordex-nonspatial.Rmd" class="external-link"><code>vignettes/concordex-nonspatial.Rmd</code></a></small>
      <div class="d-none name"><code>concordex-nonspatial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>UMAP is commonly used in scRNA-seq data analysis as a visualization
tool projecting high dimensional data onto 2 dimensions to visualize
cell clustering. However, UMAP is prone to showing spurious clustering
and distorting distances <span class="citation">(Chari, Banerjee, and
Pachter 2021)</span>. Moreover, UMAP shows clustering that seems to
correspond to graph-based clusters from Louvain and Leiden because the k
nearest neighbor graph is used in both clustering and UMAP. We have
developed <code>concordex</code> as a quantitative alternative to UMAP
cluster visualization without the misleading problems of UMAP. This
package is the R implementation of the original Python command line
tool.</p>
<p>In a nutshell, <code>concordex</code> finds the proportion of cells
among the k-nearest neighbors of each cell with the same cluster or
label as the cell itself. This is computed across all labels and the
average of all labels is returned as a metric that indicates the quality
of clustering. If the clustering separates cells well, then the observed
similarity matrix should be diagonal dominant.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/concordexR" class="external-link">concordexR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxPBMCData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>In this vignette, we demonstrate the usage of <code>concordex</code>
on a human peripheral blood mononuclear cells (PBMC) scRNA-seq dataset
from 10X Genomics. The data is loaded as a
<code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html" class="external-link">TENxPBMCData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?TENxPBMCData and browseVignettes('TENxPBMCData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span></code></pre></div>
<p>Here we plot the standard QC metrics: total number of UMIs detected
per cell (<code>nCounts</code>), number of genes detected
(<code>nGenes</code>), and percentage of UMIs from mitochondrially
encoded genes (<code>pct_mito</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">mito_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol_TENx</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">pct_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">mito_inds</span>,<span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="va">sce</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"nCounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"nGenes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"pct_mito"</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-nonspatial_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"pct_mito"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="concordex-nonspatial_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Remove the outliers and cells with high percentage of mitochondrial
counts as the high percentage is not expected biologically from the cell
type:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&lt;</span> <span class="fl">10000</span> <span class="op">&amp;</span> <span class="va">sce</span><span class="op">$</span><span class="va">pct_mito</span> <span class="op">&lt;</span> <span class="fl">8</span><span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<p>Then normalize the data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="graph-based-clustering-in-pca-space">Graph based clustering in PCA space<a class="anchor" aria-label="anchor" href="#graph-based-clustering-in-pca-space"></a>
</h2>
<p>For simplicity, the top 500 highly variable genes are used to perform
PCA:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, ntop <span class="op">=</span> <span class="fl">500</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>See the number of PCs to use later from the elbow plot:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span>, <span class="st">"percentVar"</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"Percentage of variance explained"</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-nonspatial_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Percentage of variance explained drops sharply from PC1 to PC5, and
definitely levels off after PC10, so we use the top 10 PCs for
clustering here. The graph based Leiden clustering uses a k nearest
neighbor graph. For demonstration here, we use <code>k = 10</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                           <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">10</span>, cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                        cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                          objective_function <span class="op">=</span> <span class="st">"modularity"</span></span>
<span>                                        <span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>See what the clusters look like in PCA space:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, color_by <span class="op">=</span> <span class="st">"cluster"</span>, ncomponents <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in data.frame(gg1$all, df_to_plot[, -reddim_cols]): row names were</span></span>
<span><span class="co">#&gt; found from a short variable and have been discarded</span></span></code></pre></div>
<p><img src="concordex-nonspatial_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>Some of the clusters seem well-separated along the first 4 PCs.</p>
<p>Since UMAP is commonly used to visualize the clusters, we plot UMAP
here although we don’t recommend UMAP because it’s prone to showing
spurious clusters and distorting distances. UMAP also uses a k nearest
neighbor graph, and we use the same <code>k = 10</code> here:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, n_dimred <span class="op">=</span> <span class="fl">10</span>, n_neighbors <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, color_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-nonspatial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>For the most part, the clusters are clearly separated on UMAP.</p>
</div>
<div class="section level2">
<h2 id="enter-concordex">Enter <code>concordex</code><a class="anchor" aria-label="anchor" href="#enter-concordex"></a>
</h2>
<p>Since UMAP is prone to showing spurious clusters, we’ll see what the
<code>concordex</code> metric says about the clustering and whether it
agrees with UMAP visualization. Here we explicitly obtain the k nearest
neighbor graph, as clustering and UMAP above did not store the graph
itself.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN-methods.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>The result here is a list of two <code>n</code> (number of cell) by
<code>k</code> matrices. The first is the indices of each cell’s
neighbors, as in an adjacency list that can be matrix here due to the
fixed number of neighbors, and the second is the distances between each
cell and its neighbors. For <code>concordex</code>, only the first
matrix is relevant. An adjacency matrix, either sparse of dense, as
stored in the <code>Seurat</code> object, can also be used. Here the
cluster labels are permuted 100 times.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateConcordex.html">calculateConcordex</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    labels<span class="op">=</span><span class="st">"cluster"</span>, </span>
<span>    use.dimred<span class="op">=</span><span class="st">"PCA"</span>,</span>
<span>    compute_similarity<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here the argument <code>compute_similarity</code> indicates that we
concordex will return the cluster-cluster similarity matrix. The entries
in this matrix itself represent the proportion of cells with each label
in the neighborhood of other cells with the same label.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"similarity"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">sim</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;      1    2    3    4    5    6    7    8    9</span></span>
<span><span class="co">#&gt; 1 0.71 0.00 0.00 0.00 0.03 0.00 0.24 0.01 0.00</span></span>
<span><span class="co">#&gt; 2 0.01 0.98 0.00 0.00 0.00 0.00 0.01 0.00 0.00</span></span>
<span><span class="co">#&gt; 3 0.00 0.00 0.97 0.00 0.00 0.03 0.00 0.00 0.00</span></span>
<span><span class="co">#&gt; 4 0.00 0.00 0.00 0.94 0.06 0.00 0.00 0.00 0.00</span></span>
<span><span class="co">#&gt; 5 0.13 0.00 0.00 0.02 0.76 0.00 0.05 0.03 0.00</span></span>
<span><span class="co">#&gt; 6 0.00 0.00 0.07 0.00 0.00 0.93 0.00 0.00 0.00</span></span>
<span><span class="co">#&gt; 7 0.07 0.00 0.00 0.00 0.00 0.00 0.90 0.03 0.00</span></span>
<span><span class="co">#&gt; 8 0.06 0.00 0.00 0.00 0.02 0.00 0.50 0.42 0.00</span></span>
<span><span class="co">#&gt; 9 0.00 0.00 0.28 0.29 0.00 0.08 0.00 0.00 0.33</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.2.0             scater_1.32.0              </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.1               scuttle_1.14.0             </span></span>
<span><span class="co">#&gt;  [5] bluster_1.14.0              BiocNeighbors_1.22.0       </span></span>
<span><span class="co">#&gt;  [7] TENxPBMCData_1.22.0         HDF5Array_1.32.0           </span></span>
<span><span class="co">#&gt;  [9] rhdf5_2.48.0                DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt; [11] SparseArray_1.4.8           S4Arrays_1.4.1             </span></span>
<span><span class="co">#&gt; [13] abind_1.4-5                 Matrix_1.7-0               </span></span>
<span><span class="co">#&gt; [15] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [17] Biobase_2.64.0              GenomicRanges_1.56.1       </span></span>
<span><span class="co">#&gt; [19] GenomeInfoDb_1.40.1         IRanges_2.38.1             </span></span>
<span><span class="co">#&gt; [21] S4Vectors_0.42.1            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt; [23] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt; [25] concordexR_0.99.3           BiocStyle_2.32.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.3                 gridExtra_2.3            </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.4               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] compiler_4.4.1            RSQLite_2.3.7            </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.26.0 png_0.1-8                </span></span>
<span><span class="co">#&gt;   [9] systemfonts_1.1.0         vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [11] pkgconfig_2.0.3           SpatialExperiment_1.14.0 </span></span>
<span><span class="co">#&gt;  [13] crayon_1.5.3              fastmap_1.2.0            </span></span>
<span><span class="co">#&gt;  [15] dbplyr_2.5.0              magick_2.8.4             </span></span>
<span><span class="co">#&gt;  [17] XVector_0.44.0            labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [19] utf8_1.2.4                rmarkdown_2.27           </span></span>
<span><span class="co">#&gt;  [21] ggbeeswarm_0.7.2          UCSC.utils_1.0.0         </span></span>
<span><span class="co">#&gt;  [23] ragg_1.3.2                purrr_1.0.2              </span></span>
<span><span class="co">#&gt;  [25] bit_4.0.5                 xfun_0.46                </span></span>
<span><span class="co">#&gt;  [27] zlibbioc_1.50.0           cachem_1.1.0             </span></span>
<span><span class="co">#&gt;  [29] beachmat_2.20.0           jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt;  [31] blob_1.2.4                highr_0.11               </span></span>
<span><span class="co">#&gt;  [33] rhdf5filters_1.16.0       Rhdf5lib_1.26.0          </span></span>
<span><span class="co">#&gt;  [35] BiocParallel_1.38.0       irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [37] parallel_4.4.1            cluster_2.1.6            </span></span>
<span><span class="co">#&gt;  [39] R6_2.5.1                  bslib_0.7.0              </span></span>
<span><span class="co">#&gt;  [41] jquerylib_0.1.4           Rcpp_1.0.13              </span></span>
<span><span class="co">#&gt;  [43] bookdown_0.40             knitr_1.48               </span></span>
<span><span class="co">#&gt;  [45] FNN_1.1.4                 igraph_2.0.3             </span></span>
<span><span class="co">#&gt;  [47] tidyselect_1.2.1          viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [49] yaml_2.3.9                codetools_0.2-20         </span></span>
<span><span class="co">#&gt;  [51] curl_5.2.1                lattice_0.22-6           </span></span>
<span><span class="co">#&gt;  [53] tibble_3.2.1              withr_3.0.0              </span></span>
<span><span class="co">#&gt;  [55] KEGGREST_1.44.1           evaluate_0.24.0          </span></span>
<span><span class="co">#&gt;  [57] desc_1.4.3                isoband_0.2.7            </span></span>
<span><span class="co">#&gt;  [59] BiocFileCache_2.12.0      ExperimentHub_2.12.0     </span></span>
<span><span class="co">#&gt;  [61] Biostrings_2.72.1         pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [63] BiocManager_1.30.23       filelock_1.0.3           </span></span>
<span><span class="co">#&gt;  [65] generics_0.1.3            BiocVersion_3.19.1       </span></span>
<span><span class="co">#&gt;  [67] sparseMatrixStats_1.16.0  munsell_0.5.1            </span></span>
<span><span class="co">#&gt;  [69] scales_1.3.0              glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [71] tools_4.4.1               AnnotationHub_3.12.0     </span></span>
<span><span class="co">#&gt;  [73] ScaledMatrix_1.12.0       fs_1.6.4                 </span></span>
<span><span class="co">#&gt;  [75] grid_4.4.1                colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [77] AnnotationDbi_1.66.0      GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">#&gt;  [79] beeswarm_0.4.0            BiocSingular_1.20.0      </span></span>
<span><span class="co">#&gt;  [81] vipor_0.4.7               rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [83] cli_3.6.3                 rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [85] textshaping_0.4.0         fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [87] viridisLite_0.4.2         dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [89] uwot_0.2.2                gtable_0.3.5             </span></span>
<span><span class="co">#&gt;  [91] sass_0.4.9                digest_0.6.36            </span></span>
<span><span class="co">#&gt;  [93] ggrepel_0.9.5             farver_2.1.2             </span></span>
<span><span class="co">#&gt;  [95] rjson_0.2.21              htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [97] memoise_2.0.1             htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [99] pkgdown_2.1.0             lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt; [101] httr_1.4.7                mime_0.12                </span></span>
<span><span class="co">#&gt; [103] MASS_7.3-60.2             bit64_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Chari2021-hb" class="csl-entry">
Chari, Tara, Joeyta Banerjee, and Lior Pachter. 2021. <span>“The
Specious Art of <span>Single-Cell</span> Genomics.”</span>
<em>bioRxiv</em>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kayla Jackson, A. Sina Booeshaghi, Angel Galvez-Merchan, Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
