<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="concordexR">
<title>Using concordex in place of UMAP in scRNA-seq • concordexR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using concordex in place of UMAP in scRNA-seq">
<meta property="og:description" content="concordexR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">concordexR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/concordex-demo.html">Using concordex in place of UMAP in scRNA-seq</a>
    <a class="dropdown-item" href="../articles/overview.html">Overview of concordexR</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/concordexR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Using concordex in place of UMAP in scRNA-seq</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">Apr 03, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/concordexR/blob/HEAD/vignettes/concordex-demo.Rmd" class="external-link"><code>vignettes/concordex-demo.Rmd</code></a></small>
      <div class="d-none name"><code>concordex-demo.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>UMAP is commonly used in scRNA-seq data analysis as a visualization
tool projecting high dimensional data onto 2 dimensions to visualize
cell clustering. However, UMAP is prone to showing spurious clustering
and distorting distances <span class="citation">(Chari, Banerjee, and
Pachter 2021)</span>. Moreover, UMAP shows clustering that seems to
correspond to graph-based clusters from Louvain and Leiden because the k
nearest neighbor graph is used in both clustering and UMAP. We have
developed <code>concordex</code> as a quantitative alternative to UMAP
cluster visualization without the misleading problems of UMAP. This
package is the R implementation of the original Python command line
tool.</p>
<p>In a nutshell, <code>concordex</code> finds the proportion of cells
among the k nearest neighbors of each cell with the same cluster or
label as the cell itself. This is computed across all labels and the
average of all labels is returned as a metric that indicates the quality
of clustering. To see if this is significant, the labels are permuted to
estimate a null distribution and the actual observed value is compared
to the simulated values. If the clustering separates cells well, then
the observed value should be much higher than the simulated values,
i.e. the neighborhood of each cell is more dominated by cells of the
same label as the cell of interest than by chance.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/concordexR" class="external-link">concordexR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxPBMCData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>In this vignette, we demonstrate the usage of <code>concordex</code>
on a human peripheral blood mononuclear cells (PBMC) scRNA-seq dataset
from 10X Genomics. The data is loaded as a
<code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html" class="external-link">TENxPBMCData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-10-31</span></span>
<span><span class="co">#&gt; see ?TENxPBMCData and browseVignettes('TENxPBMCData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span></code></pre></div>
<p>Here we plot the standard QC metrics: total number of UMIs detected
per cell (<code>nCounts</code>), number of genes detected
(<code>nGenes</code>), and percentage of UMIs from mitochondrially
encoded genes (<code>pct_mito</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">mito_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol_TENx</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">pct_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">mito_inds</span>,<span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="va">sce</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"nCounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"nGenes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"pct_mito"</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sce</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"pct_mito"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Remove the outliers and cells with high percentage of mitochondrial
counts as the high percentage is not expected biologically from the cell
type:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">sce</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&lt;</span> <span class="fl">10000</span> <span class="op">&amp;</span> <span class="va">sce</span><span class="op">$</span><span class="va">pct_mito</span> <span class="op">&lt;</span> <span class="fl">8</span><span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<p>Then normalize the data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="graph-based-clustering-in-pca-space">Graph based clustering in PCA space<a class="anchor" aria-label="anchor" href="#graph-based-clustering-in-pca-space"></a>
</h2>
<p>For simplicity, the top 500 highly variable genes are used to perform
PCA:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, ntop <span class="op">=</span> <span class="fl">500</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>See the number of PCs to use later from the elbow plot:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span>, <span class="st">"percentVar"</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"Percentage of variance explained"</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Percentage of variance explained drops sharply from PC1 to PC5, and
definitely levels off after PC10, so we use the top 10 PCs for
clustering here. The graph based Leiden clustering uses a k nearest
neighbor graph. For demonstration here, we use <code>k = 10</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                           <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">10</span>, cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                        cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                          objective_function <span class="op">=</span> <span class="st">"modularity"</span></span>
<span>                                        <span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>See what the clusters look like in PCA space:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, color_by <span class="op">=</span> <span class="st">"cluster"</span>, ncomponents <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in data.frame(gg1$all, df_to_plot[, -reddim_cols]): row names were</span></span>
<span><span class="co">#&gt; found from a short variable and have been discarded</span></span>
<span><span class="co">#&gt; Warning: The dot-dot notation (`..scaled..`) was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `after_stat(scaled)` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">scater</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://support.bioconductor.org/&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>Some of the clusters seem well-separated along the first 4 PCs.</p>
<p>Since UMAP is commonly used to visualize the clusters, we plot UMAP
here although we don’t recommend UMAP because it’s prone to showing
spurious clusters and distorting distances. UMAP also uses a k nearest
neighbor graph, and we use the same <code>k = 10</code> here:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, n_dimred <span class="op">=</span> <span class="fl">10</span>, n_neighbors <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, color_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>For the most part, the clusters are clearly separated on UMAP.</p>
</div>
<div class="section level2">
<h2 id="enter-concordex">Enter <code>concordex</code><a class="anchor" aria-label="anchor" href="#enter-concordex"></a>
</h2>
<p>Since UMAP is prone to showing spurious clusters, we’ll see what the
<code>concordex</code> metric says about the clustering and whether it
agrees with UMAP visualization. Here we explicitly obtain the k nearest
neighbor graph, as clustering and UMAP above did not store the graph
itself.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN-methods.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>The result here is a list of two <code>n</code> (number of cell) by
<code>k</code> matrices. The first is the indices of each cell’s
neighbors, as in an adjacency list that can be matrix here due to the
fixed number of neighbors, and the second is the distances between each
cell and its neighbors. For <code>concordex</code>, only the first
matrix is relevant. An adjacency matrix, either sparse of dense, as
stored in the <code>Seurat</code> object, can also be used. Here the
cluster labels are permuted 100 times.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateConcordex.html">calculateConcordex</a></span><span class="op">(</span><span class="va">g</span><span class="op">$</span><span class="va">index</span>, labels <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">cluster</span>, k <span class="op">=</span> <span class="fl">10</span>, n.iter <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>The results can be visualized in plots, which is implemented in this
R package but not in the Python package. The actual
<code>concordex</code> value can be compared to the simulated values
where the latter is visualized in a density plot:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotConcordexSim.html">plotConcordexSim</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Here the actual value is much higher than the simulated values,
indicating that the cluster labels do reflect actual clusters well. The
value itself is the proportion of cells with each label in the
neighborhood of other cells with the same label, averaged over all
labels.</p>
<p>To aid interpretation, the ratio of the observed value to the average
simulated value is also returned:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">corrected_trace</span></span>
<span><span class="co">#&gt; [1] 7.64354</span></span></code></pre></div>
<p>A number greater than 1 means that cells are more likely to have
neighbors with the same label than expected by chance when the labels
are completely randomly assigned, and the value of 7.7 here means that
the clusters are really good.</p>
<p>While the value is an average over all clusters, the matrix with the
values for each cluster is also returned by default, and can be
visualized in a heatmap, as a clustering diagnostic:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/heatConcordex.html">heatConcordex</a></span><span class="op">(</span><span class="va">res</span>, angle_col <span class="op">=</span> <span class="fl">0</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="concordex-demo_files/figure-html/unnamed-chunk-18-1.png" width="432"></p>
<p>This heatmap also indicates good clustering, as almost all neighbors
of cells from each of the cluster labels have the same label, on the
diagonal. Off diagonal entries should be interpreted as such: (i,j)
means the proportion of cells with label i in the neighborhood of cells
with label j. Off diagonal entries means ambiguity between labels as
similar cells get different labels. As overplotting easily happens on
the UMAP plot, this heatmap shows clustering quality more unambiguously
than UMAP.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.1.2             scater_1.26.1              </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.1               scuttle_1.8.4              </span></span>
<span><span class="co">#&gt;  [5] bluster_1.8.0               BiocNeighbors_1.16.0       </span></span>
<span><span class="co">#&gt;  [7] TENxPBMCData_1.16.0         HDF5Array_1.26.0           </span></span>
<span><span class="co">#&gt;  [9] rhdf5_2.42.0                DelayedArray_0.24.0        </span></span>
<span><span class="co">#&gt; [11] Matrix_1.5-3                SingleCellExperiment_1.20.1</span></span>
<span><span class="co">#&gt; [13] SummarizedExperiment_1.28.0 Biobase_2.58.0             </span></span>
<span><span class="co">#&gt; [15] GenomicRanges_1.50.2        GenomeInfoDb_1.34.9        </span></span>
<span><span class="co">#&gt; [17] IRanges_2.32.0              S4Vectors_0.36.2           </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.44.0         MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt; [21] matrixStats_0.63.0          concordexR_0.99.0          </span></span>
<span><span class="co">#&gt; [23] BiocStyle_2.26.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] AnnotationHub_3.6.0           BiocFileCache_2.6.1          </span></span>
<span><span class="co">#&gt;   [3] systemfonts_1.0.4             igraph_1.4.1                 </span></span>
<span><span class="co">#&gt;   [5] BiocParallel_1.32.6           digest_0.6.31                </span></span>
<span><span class="co">#&gt;   [7] htmltools_0.5.5               viridis_0.6.2                </span></span>
<span><span class="co">#&gt;   [9] fansi_1.0.4                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [11] memoise_2.0.1                 ScaledMatrix_1.6.0           </span></span>
<span><span class="co">#&gt;  [13] cluster_2.1.4                 Biostrings_2.66.0            </span></span>
<span><span class="co">#&gt;  [15] pkgdown_2.0.7                 colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [17] blob_1.2.4                    rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [19] ggrepel_0.9.3                 textshaping_0.3.6            </span></span>
<span><span class="co">#&gt;  [21] xfun_0.38                     dplyr_1.1.1                  </span></span>
<span><span class="co">#&gt;  [23] crayon_1.5.2                  RCurl_1.98-1.12              </span></span>
<span><span class="co">#&gt;  [25] jsonlite_1.8.4                glue_1.6.2                   </span></span>
<span><span class="co">#&gt;  [27] gtable_0.3.3                  zlibbioc_1.44.0              </span></span>
<span><span class="co">#&gt;  [29] XVector_0.38.0                BiocSingular_1.14.0          </span></span>
<span><span class="co">#&gt;  [31] Rhdf5lib_1.20.0               scales_1.2.1                 </span></span>
<span><span class="co">#&gt;  [33] pheatmap_1.0.12               DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [35] Rcpp_1.0.10                   isoband_0.2.7                </span></span>
<span><span class="co">#&gt;  [37] viridisLite_0.4.1             xtable_1.8-4                 </span></span>
<span><span class="co">#&gt;  [39] bit_4.0.5                     rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [41] httr_1.4.5                    FNN_1.1.3.2                  </span></span>
<span><span class="co">#&gt;  [43] RColorBrewer_1.1-3            ellipsis_0.3.2               </span></span>
<span><span class="co">#&gt;  [45] pkgconfig_2.0.3               farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [47] sass_0.4.5                    uwot_0.1.14                  </span></span>
<span><span class="co">#&gt;  [49] dbplyr_2.3.2                  utf8_1.2.3                   </span></span>
<span><span class="co">#&gt;  [51] tidyselect_1.2.0              labeling_0.4.2               </span></span>
<span><span class="co">#&gt;  [53] rlang_1.1.0                   later_1.3.0                  </span></span>
<span><span class="co">#&gt;  [55] AnnotationDbi_1.60.2          munsell_0.5.0                </span></span>
<span><span class="co">#&gt;  [57] BiocVersion_3.16.0            tools_4.2.3                  </span></span>
<span><span class="co">#&gt;  [59] cachem_1.0.7                  cli_3.6.1                    </span></span>
<span><span class="co">#&gt;  [61] generics_0.1.3                RSQLite_2.3.0                </span></span>
<span><span class="co">#&gt;  [63] ExperimentHub_2.6.0           evaluate_0.20                </span></span>
<span><span class="co">#&gt;  [65] stringr_1.5.0                 fastmap_1.1.1                </span></span>
<span><span class="co">#&gt;  [67] yaml_2.3.7                    ragg_1.2.5                   </span></span>
<span><span class="co">#&gt;  [69] knitr_1.42                    bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [71] fs_1.6.1                      purrr_1.0.1                  </span></span>
<span><span class="co">#&gt;  [73] KEGGREST_1.38.0               sparseMatrixStats_1.10.0     </span></span>
<span><span class="co">#&gt;  [75] mime_0.12                     compiler_4.2.3               </span></span>
<span><span class="co">#&gt;  [77] beeswarm_0.4.0                filelock_1.0.2               </span></span>
<span><span class="co">#&gt;  [79] curl_5.0.0                    png_0.1-8                    </span></span>
<span><span class="co">#&gt;  [81] interactiveDisplayBase_1.36.0 tibble_3.2.1                 </span></span>
<span><span class="co">#&gt;  [83] bslib_0.4.2                   stringi_1.7.12               </span></span>
<span><span class="co">#&gt;  [85] highr_0.10                    desc_1.4.2                   </span></span>
<span><span class="co">#&gt;  [87] lattice_0.20-45               vctrs_0.6.1                  </span></span>
<span><span class="co">#&gt;  [89] pillar_1.9.0                  lifecycle_1.0.3              </span></span>
<span><span class="co">#&gt;  [91] rhdf5filters_1.10.1           BiocManager_1.30.20          </span></span>
<span><span class="co">#&gt;  [93] jquerylib_0.1.4               bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;  [95] irlba_2.3.5.1                 httpuv_1.6.9                 </span></span>
<span><span class="co">#&gt;  [97] R6_2.5.1                      bookdown_0.33                </span></span>
<span><span class="co">#&gt;  [99] promises_1.2.0.1              gridExtra_2.3                </span></span>
<span><span class="co">#&gt; [101] vipor_0.4.5                   codetools_0.2-19             </span></span>
<span><span class="co">#&gt; [103] MASS_7.3-58.2                 rprojroot_2.0.3              </span></span>
<span><span class="co">#&gt; [105] withr_2.5.0                   GenomeInfoDbData_1.2.9       </span></span>
<span><span class="co">#&gt; [107] parallel_4.2.3                grid_4.2.3                   </span></span>
<span><span class="co">#&gt; [109] beachmat_2.14.0               rmarkdown_2.21               </span></span>
<span><span class="co">#&gt; [111] DelayedMatrixStats_1.20.0     shiny_1.7.4                  </span></span>
<span><span class="co">#&gt; [113] ggbeeswarm_0.7.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Chari2021-hb" class="csl-entry">
Chari, Tara, Joeyta Banerjee, and Lior Pachter. 2021. <span>“The
Specious Art of <span>Single-Cell</span> Genomics.”</span>
<em>bioRxiv</em>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Kayla Jackson, Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
